### PR TITLE
Updated the JavaCommon parser to deal with constructor-implementation conflicts

### PR DESCRIPTION
**Pull request by MatthewMarinets**
-Adjusted JavaCommon to deal with implementations of classes as constructor arguments.

-Added a test to JavaCommonTests to test the new functionality.

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `master/src/CHANGES.txt` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: src/engine/SCons/Tool/JavaCommonTests.py [hunk 1]
```diff
@@ -566,6 +566,49 @@ def test_genercis(self):
...
+package com.Matthew;
+
+public class AnonDemo {
+
+    public static void main(String[] args) {
+        new AnonDemo().execute();
+    }
+
+    public void execute() {
+        Foo bar = new Foo(new Foo() {
```
Comment (bdbaddog) [inline src/engine/SCons/Tool/JavaCommonTests.py:584]: Is this the  construct which was previously not being processed properly?

**Comment by bdbaddog** [comment]
Are these bugs related:
https://github.com/SCons/scons/issues/2771
https://github.com/SCons/scons/issues/2742
https://github.com/SCons/scons/issues/2548
https://github.com/SCons/scons/issues/2359
https://github.com/SCons/scons/issues/1849
https://github.com/SCons/scons/issues/1594

If any or all are fixed by your changes it'd great to mark these issues closed.
(apologies for a long list but at quick glance they seemed possibly related)

**Comment by MatthewMarinets** [comment]
Keep in mind that when I look at the bugs, I can only really comment on JDK 1.8. I'm still fairly new to Java and don't have the means or the know-how to test different Java versions.
#2771 
I **can't reproduce the issue**. With both test cases, the output.class files match between what javac outputs, what my version of the Java() emitter outputs, and what the old version of the Java() emitter outputs. At least for JDK 1.8, this doesn't seem to be an issue at all.
actual = new = old
#2742 
I didn't look at this one too hard; it had more to do with the Jar() builder than the Java() builder from what I can understand. **Mostly unrelated**
For most of my company's build I use a Jar-based workaround to Java emitter problems. I get all my class files in one build action, put them in the same directory, and jar the entire directory. I add an explicit dependency of the jar on the sources, so whenever a source file changes, javac kicks in and its all rejarred even if the only change was in class files SCons couldn't find. This workaround is still necessary, because I'm aware that there are still a few anonymous inner class files that SCons doesn't pick up, but I don't know what causes that. The main purpose of my change was to avoid having SCons expect files that would never be build, as that would cause unnecessary rebuilds and I couldn't find any other workaround for that.
Fixed #2548 
This seems to be exactly the bug I ran into. This one is the one that's **fixed**.
Just to be sure I ran the example test case in the post. actual = new != old
Fixed #2359
This seems to be a duplicate of #2548 . **Fixed**
actual = new != old
#1849
Bug seems unrelated, I **couldn't reproduce the issue**
actual = new = old
As for the patch, I'm pretty sure I touched some of the code mentioned in the comments thread. In particular, line 246 of my new JavaCommon.py is what does it:
"self.outer_state._getAnonStack().pop()"
I had to move around the management of the anonStack to get it to work properly. Essentially, the old version would only do any kind of stack management if it was in OuterState; if it smelled like an anon class was coming, it would add it to the stack and wait until it was sure about the anon class to pop it back off. This caused a bug because it was possible to queue up 2 or more anon classes within a constructor, and it would only pop the stack once, locking you into having an extra item in your stack. I fixed this by letting ScopeState pop items off the anon class stack whenever a ScopeState was popped off the parser state stack. Hence, these anon classes could be queued up properly.
TL;DR: The proposed solution in this bug report caused another, more nuanced bug. I fixed it (I think).
#1594 
This one is **still broken**. This might actually be the minimal case for one of the other issues I was never able to find.
actual != (new = old)
Sidenote: I agree with electricalan in this thread, that SCons really isn't great at languages with unpredictable intermediate files. I outlined the workaround I've been using above in my response to #2742 , and I've found another workaround using the inline action execs with "${}". It's doable, easy even with the right helper functions, but it isn't pretty.
Or maybe it is, and I just never found out how to do it.

Sorry for the long response time, I had the page open and didn't notice that I needed to refresh.
-Matthew

**Comment by bdbaddog** [comment]
Thanks for taking the time to review all those.
Very much appreciated.
No worries about response time. We're all volunteers here. :)
